src/csharp/Grpc.Examples/MathGrpc.cs

```diff
@@ -27,10 +27,35 @@ public static partial class Math
   {
     static readonly string __ServiceName = "math.Math";
 
-    static readonly grpc::Marshaller<global::Math.DivArgs> __Marshaller_math_DivArgs = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.DivArgs.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.DivReply> __Marshaller_math_DivReply = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.DivReply.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.FibArgs> __Marshaller_math_FibArgs = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.FibArgs.Parser.ParseFrom);
-    static readonly grpc::Marshaller<global::Math.Num> __Marshaller_math_Num = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Math.Num.Parser.ParseFrom);
+    static void __Helper_WriteBufferMessage(global::Google.Protobuf.IMessage message, global::Grpc.Core.SerializationContext context)
+    {
+      #if !GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION
+      var bufferMessage = message as global::Google.Protobuf.IBufferMessage;
+      if (bufferMessage != null)
+      {
+        var writer = new global::Google.Protobuf.CodedOutputWriter(context.GetBufferWriter());
+        bufferMessage.WriteTo(ref writer);
+        writer.Flush();
+        context.Complete();
+        return;
+      }
+      #endif
+      context.Complete(global::Google.Protobuf.MessageExtensions.ToByteArray(message));
+    }
+
+    static T __Helper_ParseBufferMessage<T>(global::Grpc.Core.DeserializationContext context, global::Google.Protobuf.MessageParser<T> parser) where T : global::Google.Protobuf.IMessage<T>
+    {
+      #if !GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION
+      return parser.ParseFrom(context.PayloadAsReadOnlySequence());
```

Comment:
currently the serialization/deserialization is asymmetric in the sense that for serialization, we check if the message is IBufferMessage (and fall back if not), but for parsing we go for the ParseFrom(ROS) method directly with no fallback.
We might wanna be consistent here to avoid unexpected results.

Reply:
`ParseFrom(ROS<byte>)` does this internally. If the type doesn't implement IBufferMessage then it converts the sequence to a byte array for you.

Reply:
In those cases, wouldn't it be expected that GOOGLE_PROTOBUF_DISABLE_BUFFER_SERIALIZATION would be set to avoid the extra conversion?